[dynamo 3.11] fix segfault when printing stack trace

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99180
* __->__ #99934

Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @soumith @anijain2305 @Xia-Weiwen @desertfire

williamwen42 committed:
[dynamo 3.11] fix segfault when printing stack trace

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

albanD commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -542,10 +549,20 @@ inline static PyObject* eval_custom_code(
+  result = eval_frame_default(tstate, shadow, throw_flag);
+
+  #if IS_PYTHON_3_11_PLUS
+ cleanup:
```
I know CPython uses a lot of gotos but we usually don't. There are 0 usage of goto across torch/ aten/ c10/ today.
I think it is ok to duplicate these 3 cleanup lines above in this case.

albanD commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -461,57 +461,55 @@ inline static PyObject* eval_custom_code(
+  }
+
+  size_t size = code->co_nlocalsplus + code->co_stacksize + FRAME_SPECIALS_SIZE;
+  THP_EVAL_API_FRAME_OBJECT* shadow = malloc(size);
```
I think we want a more detailed comment on why we do that and why it is ok here

albanD commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -461,57 +461,55 @@ inline static PyObject* eval_custom_code(
 
-  // copy from old fastlocals to new fastlocals:
+  for (int i = 0; i < code->co_nlocalsplus; i++) {
+    fastlocals_new[i] = NULL;
```
Why is this needed? Is that in case you don't re-populate the whole array you don't want it to be uninitialized?

williamwen42 replied:
Yes. Every element in the localsplus array will be Py_XINCREF'd.

albanD replied:
Can you have a comment here explaining that please?

albanD reviewed (commented):
Sounds pretty good.
Are there specific tests we should re-enable following this? Or add new one-off tests to show what this fixed?

williamwen42 commented:
I'll add a test that launches a subprocess running code with an unhandled exception. Other tests will be enabled in the next PR.

williamwen42 committed:
Update on "[WIP, dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

albanD commented on test/dynamo/test_misc.py:
```diff
@@ -4697,6 +4698,27 @@ def test_remove_dead_code_with_exn_table_entries(self):
+            capture_output=True,
+        )
+        # process did not segfault, C assert fail (abort), etc.
+        self.assertGreater(proc.returncode, 0)
```
Segfault would actually give you a non-zero code?

williamwen42 replied:
Exits due to signals (segfault, abort, etc.) will have negative return code.

albanD replied:
I don't think that this is generally true?
This is not true on my Fedora install at least:
```
$ python foo.py 
Segmentation fault (core dumped)
$ echo $?
139
```

williamwen42 replied:
This behavior is specific to python: https://docs.python.org/3/library/subprocess.html#subprocess.CompletedProcess.returncode

williamwen42 replied:
We'll skip the test if not on POSIX

williamwen42 replied:
Update: test changed to trigger segfault without spawning a subprocess. POSIX check no longer required.

albanD reviewed (approved):
nit for comment bug sounds good otherwise!

williamwen42 committed:
Update on "[dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

malfet commented on test/dynamo/test_misc.py:
```diff
@@ -4698,6 +4699,40 @@ def test_remove_dead_code_with_exn_table_entries(self):
+opt_fn(torch.ones(2))
+"""
+        proc = subprocess.run(
+            ["python3", "-c", code],
```
Please always use `sys.executable` as `python3` can not be the version of Python this test was executed with. Also, why not use `multiprocessing`
```suggestion
            [sys.executable, "-c", code],
```

williamwen42 committed:
Update on "[dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

malfet commented on test/dynamo/test_misc.py:
```diff
@@ -4721,6 +4722,77 @@ def test_remove_dead_code_with_exn_table_entries(self):
         self.assertEquals(tab[0].end, 4)
         self.assertEquals(tab[0].target, 6)
 
+    @unittest.skipIf(os.name != "posix", "return code check is POSIX only")
```
```suggestion
   @cpuOnly
    @unittest.skipIf(os.name != "posix", "return code check is POSIX only")
```

williamwen42 replied:
Turns out it's enough to use `traceback.format_exc()` to trigger the segfault - no need to spawn a separate process.

malfet reviewed (approved):
Can you please gate those checks to CPU only shards, as forking on GPUs runner is prohibitively expensive, but the crash is reproducible on CPU machines, isn't it?)

williamwen42 committed:
Update on "[dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo 3.11] fix segfault when printing stack trace"


Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 commented:
Dynamo-wrapped pytorch tests look good for the most part at the moment up the stack (https://github.com/pytorch/pytorch/pull/99180). Will now merge this PR.

williamwen42 commented:
@pytorchbot merge
